engine-nextgen/src/main/kotlin/graphql/nadel/validation/NadelHydrationValidation.kt

```diff
@@ -91,40 +74,44 @@ internal class NadelHydrationValidation(
     private fun getOutputTypeIssues(
         parent: NadelServiceSchemaElement,
         overallField: GraphQLFieldDefinition,
-        actorService: Service,
         actorField: GraphQLFieldDefinition,
-        isPolymorphicHydration: Boolean,
+        hasMoreThanOneHydration: Boolean,
     ): List<NadelSchemaValidationError> {
         // Ensures that the underlying type of the actor field matches with the expected overall output type
-        var overallType = overallField.type.unwrapAll()
-        if (isPolymorphicHydration) {
-            if (overallType is GraphQLUnionType) {
-                val possibleObjectTypes = overallType.types
-                val actorFieldReturnType = actorField.type.unwrapAll().name
-                val overallTypeMatchingActorFieldReturnType = possibleObjectTypes.find {
-                    actorFieldReturnType == getUnderlyingType(it, actorService)?.name
+        val overallType = overallField.type.unwrapAll()
+
+        // Polymorphic hydration must have union output
+        if (hasMoreThanOneHydration && overallType !is GraphQLUnionType) {
+            return listOf(FieldWithPolymorphicHydrationMustReturnAUnion(parent, overallField))
+        }
+
+        val acceptableOutputTypes: List<GraphQLNamedOutputType> = when (overallType) {
+            is GraphQLUnionType -> overallType.types + overallType
+            is GraphQLInterfaceType -> overallSchema.getImplementations(overallType) + overallType
+            else -> listOf(overallType as GraphQLNamedOutputType)
+        }
+
+        val actorOutputTypes: List<GraphQLNamedOutputType> = when (val actorOutputType = actorField.type.unwrapAll()) {
+            is GraphQLUnionType -> actorOutputType.types
+            is GraphQLInterfaceType -> overallSchema.getImplementations(actorOutputType)
+            else -> listOf(actorOutputType as GraphQLNamedOutputType)
+        }
+
+        val typeValidation = actorOutputTypes
+            // Find incompatible output types
+            .filter { actorOutputType ->
+                acceptableOutputTypes.none { acceptableOutputType ->
+                    typeValidation.isAssignableTo(lhs = acceptableOutputType, rhs = actorOutputType)
                 }
-                val overallTypeName = overallTypeMatchingActorFieldReturnType?.name
-                    ?: return listOf(
-                        PolymorphicHydrationReturnTypeMismatch(
-                            actorField,
-                            actorService,
-                            parent,
-                            overallField
-                        )
-                    )
-                overallType = overallSchema.getType(overallTypeName) as GraphQLUnmodifiedType
-            } else {
-                return listOf(FieldWithPolymorphicHydrationMustReturnAUnion(parent, overallField))
             }
-        }
-        val typeValidation = typeValidation.validate(
```

Comment:
No need to run type validation here as we no longer deal with two type where one is underlying and one is overall.

The validation for the types is already done elsewhere. So in the code here, if they're assignable to each other, then the types are valid.